module/support/optimisation/NSGA2Evaluator/src/tasks/RotationEvaluator.cpp

```diff
@@ -0,0 +1,237 @@
+        // omega           = sensors.gyroscope.z();
+        // deltaT          = sim_time - old_time;
+        // old_time        = sim_time;
+        // theta += omega * deltaT / 1000;
+        update_max_field_plane_sway(sensors);
+
+        // Transform to torso {t} from world {w} space
+        Eigen::Matrix4d Hwt = sensors.Htw.inverse().matrix();
+        // Basis Z vector of torso {t} in world {w} space
+        Eigen::Vector3d uZTw = Hwt.block(0, 2, 3, 1);
+
+        // Check if angle between torso z axis and world z axis is greater than config value cfg.fallen_angle
+        if (!fallen && std::acos(Eigen::Vector3d::UnitZ().dot(uZTw)) > cfg.fallen_angle) {
+            NUClear::log<NUClear::DEBUG>("Fallen!");
+            fallen = true;
+            return true;
+        }
+        return false;
+    }
+
+    void RotationEvaluator::update_max_field_plane_sway(const Sensors& sensors) {
+        auto accelerometer = sensors.accelerometer;
+
+        // Calculate the robot sway along the field plane (left/right, forward/backward)
+        double field_plane_sway = std::pow(std::pow(accelerometer.x(), 2) + std::pow(accelerometer.y(), 2), 0.5);
+        if (field_plane_sway > max_field_plane_sway) {
+            max_field_plane_sway = field_plane_sway;
+        }
+    }
+
+    std::unique_ptr<NSGA2FitnessScores> RotationEvaluator::calculate_fitness_scores(bool early_termination,
+                                                                                    int generation,
+                                                                                    int individual) {
+        double trial_duration =
+            std::chrono::duration_cast<std::chrono::milliseconds>(NUClear::clock::now() - trial_start_time).count();
+
+        auto scores      = calculate_scores();
+        auto constraints = early_termination ? calculate_constraints(trial_duration) : constraints_not_violated();
+
+        NUClear::log<NUClear::DEBUG>("Trial ran for", trial_duration);
+        NUClear::log<NUClear::DEBUG>("SendFitnessScores for generation", generation, "individual", individual);
+        NUClear::log<NUClear::DEBUG>("    scores:", scores[0], scores[1]);
+        NUClear::log<NUClear::DEBUG>("    constraints:", constraints[0], constraints[1]);
+
+        // Create the fitness scores message based on the given results and emit it back to the Optimiser
+        std::unique_ptr<NSGA2FitnessScores> fitness_scores = std::make_unique<NSGA2FitnessScores>();
+        fitness_scores->id                                 = individual;
+        fitness_scores->generation                         = generation;
+        fitness_scores->obj_score                          = scores;
+        fitness_scores->constraints                        = constraints;
+        return fitness_scores;
+    }
+
+    std::vector<double> RotationEvaluator::calculate_scores() {
+
+        return {
+            max_field_plane_sway,  // For now, we want to reduce this
```

Comment:
Should there be a TODO here somewhere for this?

Reply:
hmm the commenting is maybe hard to follow here. the idea of tracking the maximum sway is to feed it into the evaluator. NSGA2 is specifically for optimising multiple values, and torso sway was one of the values optimised along with distance. The comment is probably eluding to the fact that reducing this value might not be necessary for a good walk but they just needed to come up with a second value to optimise. I think it's an okay metric and probably is too high level to make a todo from, but I'll have a look at the comments and see if anything can be clarified.